### DISCUSSION TITLE
how to set special_collision_constraint ?

### DISCUSSION THREAD

**Discussion by vegfood** [tesseract-robotics/tesseract#811] [Q&A]
auto cnt_safety_margin = std::make_shared<util::SafetyMarginData>(-0.1, 20);
        cnt_safety_margin->setPairSafetyMarginData("forearm_link", "wrist_2_link", -0.1, 20);
        composite_profile->special_collision_constraint = cnt_safety_margin;

Q: I wnat to set negative safety margin for the above two links like this but it doesn't work

**Comment by Levi-Armstrong** [comment]
That should work, but you may be running into an edge case because I have never tested a negative contact margin that large. Though I want to check because this will allow those two links to intersect by 10cm; is that what you want?

**Reply by vegfood** [reply]
10 cm is just a test, i find it reports that dist between the two links is about -0.001 m , at first time i just set the margin = -0.01, but it doesn't work, so i choose to set a much bigger margin (10cm) to check if this work and i find the collision still exists.

**Reply by Levi-Armstrong** [reply]
Did you add this profile under all profile keys?

**Reply by vegfood** [reply]
I have only one program like this:
    CompositeInstruction program(
            "cartesian_program", CompositeInstructionOrder::ORDERED,
            ManipulatorInfo("manipulator", "agv_link", "tool0"));

    planning_server.getProfiles()->addProfile<TrajOptIfoptCompositeProfile>(
                profile_ns::TRAJOPT_IFOPT_DEFAULT_NAMESPACE, "cartesian_program", composite_profile);

i have other profiles for plan instructions like:
PlanInstruction plan_f0(target_list[0], PlanInstructionType::FREESPACE, "freespace_profile");

 do i need to add the composite profile for these profile keys?

**Reply by Levi-Armstrong** [reply]
Are you using TrajOpt or TrajoptIfopt planner?

**Reply by vegfood** [reply]
TrajoptIfopt

**Reply by Levi-Armstrong** [reply]
That planner I am not sure if it that has been verified that it is fully functioning feature.

**Reply by vegfood** [reply]
I tried the TrajOpt, it works. But i wanna know if this the collision body used for planning, it seems it's much bigger than the original link, perhaps this is why there is always collision reports. Is it possible to reduce the collision body 
![Screenshot from 2022-08-17 09-27-14](https://user-images.githubusercontent.com/110875386/185015052-af5b768b-666f-4d3c-88d8-93caa2589588.png)
?

**Reply by Levi-Armstrong** [reply]
If you are using one of the Tesseract visualization widgets and you enable the visualize the collision it will show the exact collision object used for collision checking. Also the latest Tesseract Workbench plugin allows you to query collision and will show you where the collision points are in rviz.

**Reply by vegfood** [reply]
thanks for your help,  i find it

**Reply by Levi-Armstrong** [reply]
I looked at TrajOptIfopt and the special safety margins have not been ported, but if you needed to use TrajOptIfopt it should be easy for you to port. If you end up going this route please create a PR and I will happily  review and get it merged in.
